platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -13,204 +13,259 @@
    *
-   * <p>[Y, Y] (4) Felix file prop did not change, do nothing. This is the happy path.
+   * <p>[Y, null] [Y, X] (3) Felix prop was externally changed or removed. This is an error.
    *
-   * @param event the configuration event for which an update is needed
+   * <p>[Y, Y] (4) Felix file prop did not change, do nothing. This is the happy path. Properties
+   * will be updated on disk if necessary.
    */
-  private void handleUpdate(ConfigurationEvent event) {
-    Configuration configuration = getConfiguration(event);
-    if (configuration == null) {
-      LOGGER.debug("Configuration was null for pid {}", event.getPid());
-      return;
-    }
+  @Override
+  public void handleStore(ConfigurationContext context) throws IOException {
+    String pid = context.getServicePid();
+    File fileFromConfig = context.getConfigFile();
 
-    FelixConfig felixConfig = new FelixConfig(configuration);
-    String pid = felixConfig.getPid();
-    File felixFileNew = felixConfig.getFelixFile();
-    File felixFileOriginal = pidFileMap.get(pid);
+    CachedConfigData cachedConfigData = pidDataMap.get(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
 
-    // Config not tracked and will not be tracked? (1)
-    if (felixFileOriginal == null && felixFileNew == null) {
+    if (fileFromCache == null && fileFromConfig == null) {
+      // This config doesn't have an etc file, so we ignore this case (1)
       LOGGER.debug("Was not tracked and will not track pid {}", pid);
       return;
     }
 
-    boolean filePropChanged = felixConfig.filePropChanged(felixFileOriginal);
-
-    // Felix file prop did not change? (4)
-    if (filePropChanged) {
-      // Are we seeing the config for the first time? (2) (3)
-      if (felixFileOriginal != null) {
-        try {
-          felixConfig.setFelixFile(felixFileOriginal);
-        } catch (IOException e) {
-          LOGGER.error("Could not set felix file name, error writing to config admin: ", e);
-        }
-        return;
-      }
-      pidFileMap.put(pid, felixFileNew);
+    if (fileFromCache == null) {
+      // An etc config file was just dropped and we're seeing it for the first time (2)
       LOGGER.debug("Tracking pid {}", pid);
+      CachedConfigData createdConfigData = new CachedConfigData(context);
+      pidDataMap.put(pid, createdConfigData);
+      writeIfNecessary(context, createdConfigData);
+      return;
+    }
+
+    if (!Objects.equals(fileFromConfig, fileFromCache)) {
+      // The felix file prop changed, which is not allowed (3)
+      String msg =
+          (fileFromConfig == null)
+              ? format("Felix filename has been illegally removed, was [%s]", fileFromCache)
+              : format(
+                  "Felix filename has been illegally changed from [%s] to [%s]",
+                  fileFromCache, fileFromConfig);
+      throw new IllegalStateException(msg);
```

Comment:
✏️ This doesn't feel like an `IOException` and that's the only documented exception. Would definitely be clear enough to add this rt exception to the documentation of the interface.

Reply:
We cannot suddenly use another runtime exception since this exception will bubble out of the config admin API. IOException is the only thing we can use without breaking the felix API and we cannot re-document a known API that is out of our control.

Reply:
I think we are in luck. Updates are already documented to throw illegal state. We might be able to get away with this on the argument that, if you change the config file (a felix internal, nothing to do with osgi) we interpret that as the config disappearing / being deleted: 
https://osgi.org/javadoc/r5/enterprise/org/osgi/service/cm/Configuration.html#update(java.util.Dictionary)

I think this needs to be an illegal state exception. It will not come up in production unless an integrator has mucked with things incorrectly. If they do not tamper with Felix internals in the first place, they should never see this.

Reply:
I just wanted to make sure that it was a documented exception of the API which it is. Based on what I am reading, I thing IllegalArgumentException is more appropriate than IllegalStateException.